### PR TITLE
DynComms Development Branch

### PR DESCRIPTION
**Pull request by ProofOfKeags**
## Change Description
~~Depends on #8270.~~
Depends on #8981

This is a series of single responsibility commits that improve the state of affairs within the channel state machine. The ultimate goal here in service of the Dynamic Commitments project is to make it such that we no longer have the `updateLog`s contain `paymentDescriptor`s as the fundamentally tracked unit of update. This is unfortunately a long road which requires a number of changes which on their face do not *seem* necessary. I assure you that every one of these commits is ultimately in service of that goal, though.

It is highly recommended that you review commits one by one and in order. One of the consequences of having atomic, semantics-invariant commits is that some subsequent commits may ultimately make prior commits partially or fully obsolete. This is the price of making commit-by-commit review easy.

There is no specific requirement that every commit in this PR go in at once. This PR is mergeable at every commit and will be semantics-invariant.

Godspeed 🫡

## Steps to Test
make unit pkg=lnwallet

### Testing
- [ ] Your PR passes all CI checks.
- [ ] Tests covering the positive and negative (error paths) are included.
- [ ] Bug fixes contain tests triggering the bug to prevent regressions.

### Code Style and Documentation
- [ ] The change obeys the [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation) guidelines, and lines wrap at 80.
- [ ] Commits follow the [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure).
- [ ] Any new logging statements use an appropriate subsystem and logging level.
- [ ] Any new lncli commands have appropriate tags in the comments for the rpc in the proto file.
- [ ]  [There is a change description in the release notes](https://github.com/lightningnetwork/lnd/tree/master/docs/release-notes), or `[skip ci]` in the commit message for small changes.

📝 Please see our  [Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) for further guidance.

### CODE REVIEW

File: lnwallet/channel.go [hunk 1]
```diff
@@ -2709,8 +2753,26 @@ func (lc *LightningChannel) evaluateHTLCView(view *htlcView, ourBalance,
 			continue
 		}
 
-		processAddEntry(entry, ourBalance, theirBalance, nextHeight,
-			whoseCommitChain, true, mutateState)
+		// Skip the entries that have already had their add commit
+		// height set for this commit chain.
+		addHeights := &entry.addCommitHeights
+		addHeight := addHeights.GetForParty(whoseCommitChain)
+		if addHeight == 0 {
```
Comment (yyforyongyu) [inline lnwallet/channel.go]: Looks like the `Add/Remove/FeeUpdate` have the common logic,
- we get the value based on `whoseCommitChain`
- we decide to process it or not based on its value == 0 or not.

Can be simplified?

File: lnwallet/channel.go [hunk 2]
```diff
@@ -2629,9 +2629,11 @@ func (lc *LightningChannel) evaluateHTLCView(view *htlcView, ourBalance,
 			)
 
 			if h == 0 {
-				processFeeUpdate(
-					entry, &newView.feePerKw, nextHeight,
-					whoseCommitChain,
+				// If the update wasn't already locked in,
+				// update the current fee rate to reflect this
+				// update.
+				newView.feePerKw = chainfee.SatPerKWeight(
```
Comment (yyforyongyu) [inline lnwallet/channel.go]: 👍

File: lnwallet/channel_test.go [hunk 3]
```diff
@@ -8658,223 +8658,6 @@ type heights struct {
 	remoteRemove uint64
 }
 
-// TestProcessFeeUpdate tests the applying of fee updates and mutation of
```
Comment (yyforyongyu) [inline lnwallet/channel_test.go:8609]: hmm but we need to test it tho

File: lnwallet/channel.go [hunk 4]
```diff
@@ -2635,35 +2633,16 @@ func (lc *LightningChannel) evaluateHTLCView(view *htlcView, ourBalance,
 				newView.feePerKw = chainfee.SatPerKWeight(
 					entry.Amount.ToSatoshis(),
 				)
-
```
Comment (yyforyongyu) [inline lnwallet/channel.go]: 🎉🎉🎉

File: lnwallet/channel.go [hunk 5]
```diff
@@ -2798,21 +2740,50 @@ func (lc *LightningChannel) evaluateHTLCView(view *htlcView, ourBalance,
...
 		}
 
 		newView.theirUpdates = append(newView.theirUpdates, entry)
 	}
 
-	return newView, nil
+	// Create a function that is capable of identifying whether or not the
+	// paymentDescriptor has been committed in the commitment chain
+	// corresponding to whoseCommitmentChain.
+	isUncommitted := func(update *paymentDescriptor) bool {
```
Comment (yyforyongyu) [inline lnwallet/channel.go:2729]: make this a method and test it?

File: lnwallet/channel.go [hunk 6]
```diff
@@ -4191,11 +4162,59 @@ func (lc *LightningChannel) computeView(view *htlcView,
...
-	filteredHTLCView, err := lc.evaluateHTLCView(view, &ourBalance,
-		&theirBalance, nextHeight, whoseCommitChain, updateState)
+	filteredHTLCView, uncommitted, err := lc.evaluateHTLCView(
+		view, &ourBalance, &theirBalance, whoseCommitChain,
+	)
 	if err != nil {
 		return 0, 0, 0, nil, err
 	}
+	if updateState {
+		state := lc.channelState
```
Comment (yyforyongyu) [inline lnwallet/channel.go:4115]: this should be in its own method

File: lnwallet/channel.go [hunk 7]
```diff
@@ -2835,10 +2835,9 @@ func (lc *LightningChannel) fetchParent(entry *paymentDescriptor,
 // was committed is updated. Keeping track of this inclusion height allows us to
 // later compact the log once the change is fully committed in both chains.
 func processAddEntry(htlc *paymentDescriptor, ourBalance,
-	theirBalance *lnwire.MilliSatoshi, _ lntypes.ChannelParty,
-	isIncoming bool) {
+	theirBalance *lnwire.MilliSatoshi, _, originator lntypes.ChannelParty) {
 
-	if isIncoming {
+	if originator == lntypes.Remote {
```
Comment (yyforyongyu) [inline lnwallet/channel.go]: I think `isIncoming` is easier to understand - when I see `isIncoming` I know it's an incoming HTLC, when I see the originator is the remote I have to think about it.

File: lnwallet/channel.go [hunk 8]
```diff
@@ -2604,6 +2604,18 @@ func (lc *LightningChannel) evaluateHTLCView(view *htlcView, ourBalance,
...
 	}
 
+	// The fee rate of our view is always the last UpdateFee message from
+	// the channel's OpeningParty.
+	openerUpdates := view.updates.GetForParty(lc.channelState.Initiator())
+	feeUpdates := fn.Filter(func(u *paymentDescriptor) bool {
+		return u.EntryType == FeeUpdate
+	}, openerUpdates)
+	if len(feeUpdates) > 0 {
+		newView.feePerKw = chainfee.SatPerKWeight(
```
Comment (yyforyongyu) [inline lnwallet/channel.go:2615]: where is the `h == 0` check?

File: lnwallet/channel.go [hunk 9]
```diff
@@ -4098,12 +4089,26 @@ func (lc *LightningChannel) computeView(view *htlcView,
...
 	)
 	if err != nil {
 		return 0, 0, 0, nil, err
 	}
+
+	// Add the balance deltas to the balances we got from the commitment
+	// state.
+	if deltas.Local >= 0 {
+		ourBalance += lnwire.MilliSatoshi(deltas.Local)
+	} else {
```
Comment (yyforyongyu) [inline lnwallet/channel.go:4105]: I think `ourBalance += lnwire.MilliSatoshi(deltas.Local)` is enough since `deltas.Local` can be negative.

**Review by yyforyongyu — COMMENTED** [review]
Did a quick review - have a mixed feeling now that we are pushing towards a symmetric design with the help of `Dual`, and sometimes the usage gives me a "too smart" vibe, maybe I'm being too cautious. Overall this is definitely good stuff!

File: lnwallet/channel.go (hunk 9 above)
Comment (ProofOfKeags) [reply lnwallet/channel.go:4105]: `deltas.Local` can be negative, but `MilliSatoshi` cannot, so the cast is likely to lead to undefined behavior.

File: lnwallet/channel_test.go (hunk 3 above)
Comment (ProofOfKeags) [reply lnwallet/channel_test.go:8609]: It's a one liner, does it really need a test? Note that this commit series removes a lot of the complexity that is being tested.

File: lnwallet/channel.go (hunk 7 above)
Comment (ProofOfKeags) [reply lnwallet/channel.go]: Yeah that's fair, the issue with `isIncoming` is that it promotes a boolean argument which I find doesn't fit the domain modeling as well. We can move the LinkDirection type to the lntypes package and then have the functions

```golang
func (p ChannelParty) SendDirection() LinkDirection {...}
func (p ChannelParty) RecvDirection() LinkDirection {...}
func (d LinkDirection) Origin() ChannelParty {...}
func (d LinkDirection) Target() ChannelParty {...}
```

File: lnwallet/channel.go (hunk 8 above)
Comment (ProofOfKeags) [reply lnwallet/channel.go:2615]: So the rationale for excluding it is as follows: the fee rate of an htlcView is *always the most recent fee update we have received...* irrespective of its add/remove heights.

**Comment by carlaKC** [comment]
Completed a high level pass on this, great refactor with really nice commit structure 🙏 
Noticed that a lot of this is pulled out into #9097 - is that what we should review?

**Comment by ProofOfKeags** [comment]
Closing this as it is superseded by #9158 and #9097

**Pull request closed without merging**